client/question/services/FeedbackGeneratorService.js

```diff
@@ -24,12 +24,85 @@ tie.factory('FeedbackGeneratorService', [
   'RUNTIME_ERROR_FEEDBACK_MESSAGES', 'WRONG_LANGUAGE_ERRORS', 'LANGUAGE_PYTHON',
   'CLASS_NAME_AUXILIARY_CODE', 'CLASS_NAME_SYSTEM_CODE', 'PARAGRAPH_TYPE_TEXT',
   'PARAGRAPH_TYPE_CODE', 'PARAGRAPH_TYPE_SYNTAX_ERROR',
+  'PYTHON_PRIMER_BUTTON_NAME', 'ERROR_COUNTER_LANGUAGE_UNFAMILIARITY',
+  'ERROR_COUNTER_SAME_RUNTIME', 'UNFAMILIARITY_THRESHOLD',
   function(
     FeedbackObjectFactory, TranscriptService, ReinforcementGeneratorService,
     CODE_EXECUTION_TIMEOUT_SECONDS, SUPPORTED_PYTHON_LIBS,
     RUNTIME_ERROR_FEEDBACK_MESSAGES, WRONG_LANGUAGE_ERRORS, LANGUAGE_PYTHON,
     CLASS_NAME_AUXILIARY_CODE, CLASS_NAME_SYSTEM_CODE, PARAGRAPH_TYPE_TEXT,
-    PARAGRAPH_TYPE_CODE, PARAGRAPH_TYPE_SYNTAX_ERROR) {
+    PARAGRAPH_TYPE_CODE, PARAGRAPH_TYPE_SYNTAX_ERROR,
+    PYTHON_PRIMER_BUTTON_NAME, ERROR_COUNTER_LANGUAGE_UNFAMILIARITY,
+    ERROR_COUNTER_SAME_RUNTIME, UNFAMILIARITY_THRESHOLD) {
+
+    /**
+     * Counter to keep track of language unfamiliarity errors which include
+     * syntax errors and wrong language errors.
+     *
+     * @type {number}
+     */
+    var consecutiveLanguageUnfamiliarityCounter = 0;
+
+    /**
+     * Counter to keep track of consecutive same errors.
+     *
+     * @type {number}
+     */
+    var consecutiveSameRuntimeErrorCounter = 0;
+
+    /**
+     * Variable to store the error string immediately before the current error.
+     * Will be used to see if the user is receiving the same exact error
+     * consecutively, a possible indication of language unfamiliarity.
+     *
+     * @type {string}
+     */
+    var previousErrorString = '';
+
+    /**
+     * Updates the counter specified in the paramter to track language
+     * unfamiliarity and resets all other counters.
+     *
+     * @param {string} counterToIncrement
+     * @private
+     */
+    var _updateCounters = function(counterToIncrement) {
```

Comment:
You are passing in a string here, but in some cases you call this method with `undefined` as an arg. I suggest that, in those cases, you call a separate _resetCounters() method instead that takes no args, and have this function assume in all cases that its input is a string.